Tests/src/PublicAPITests/SpaceSystemTests.cpp

```diff
@@ -2853,4 +2853,75 @@ CSP_PUBLIC_TEST(CSPEngine, SpaceSystemTests, DuplicateSpaceTest)
 	// Log out
 	LogOut(UserSystem);
 }
+#endif
+
+#if RUN_ALL_UNIT_TESTS || RUN_SPACESYSTEM_TESTS || RUN_SPACESYSTEM_LEAVESCOPES_TEST
+CSP_PUBLIC_TEST(CSPEngine, SpaceSystemTests, LeaveScopesTest)
+{
+	SetRandSeed();
+
+	auto& SystemsManager = ::SystemsManager::Get();
+	auto* UserSystem	 = SystemsManager.GetUserSystem();
+	auto* SpaceSystem	 = SystemsManager.GetSpaceSystem();
+	auto* Connection	 = SystemsManager.GetMultiplayerConnection();
+
+	const char* TestSpaceName		 = "CSP-TEST-SPACE";
+	const char* TestSpaceDescription = "CSP-TEST-SPACEDESC";
+
+	char UniqueSpaceName1[256], UniqueSpaceName2[256];
+	SPRINTF(UniqueSpaceName1, "%s-%s", TestSpaceName, GetUniqueString().c_str());
+	SPRINTF(UniqueSpaceName2, "%s-%s", TestSpaceName, GetUniqueString().c_str());
+
+	String UserId;
+
+	// Log in
+	LogIn(UserSystem, UserId);
+
+	// Create space
+	Array<InviteUserRoleInfo> UserRoles(1);
+	UserRoles[0].UserEmail = AlternativeLoginEmail;
+	UserRoles[0].UserRole  = SpaceUserRole::User;
+	InviteUserRoleInfoCollection InviteInfo;
+	InviteInfo.InviteUserRoleInfos = UserRoles;
+
+	::Space Space1, Space2;
+	CreateSpace(SpaceSystem, UniqueSpaceName1, TestSpaceDescription, SpaceAttributes::Private, nullptr, InviteInfo, nullptr, nullptr, Space1);
+	CreateSpace(SpaceSystem, UniqueSpaceName2, TestSpaceDescription, SpaceAttributes::Private, nullptr, InviteInfo, nullptr, nullptr, Space2);
+
+	// Setup Leave Scopes callback
+	bool LeaveScopesCallbackCalled = false;
+
+	auto LeaveScopesCallback = [&LeaveScopesCallbackCalled](const csp::common::Array<csp::common::String>&, const csp::common::String&)
+	{
+		if (LeaveScopesCallbackCalled)
+		{
+			return;
+		}
+
+		LeaveScopesCallbackCalled = true;
+	};
+
+	Connection->SetLeaveScopesCallback(LeaveScopesCallback);
+
+	// Enter space (i.e. enter scopes)
+	auto [EnterSpace1Result] = AWAIT_PRE(SpaceSystem, EnterSpace, RequestPredicate, Space1.Id);
+
+	// Exit space (i.e. leave scopes)
+	auto [ExitSpaceResult] = AWAIT_PRE(SpaceSystem, ExitSpace, RequestPredicate);
```

Comment:
I added the ExitSpace await, but the LeaveScopes callback was not called, so then I experimented with the below (re-enter the same space without exiting first, enter a different space without exiting the first one first), but failed to receive the LeaveScopes callback. Then I realised that SetScopes and ResetScopes and the likes actually "Invoke" commands, so I tried to do the same thing (see another comment), but am unsure what method to invoke.

- Do I need to explicitly invoke LeaveScopes inside ExitSpace on the same model as ResetScopes, if so which method do I have to invoke?
- If not, how do I trigger the LeaveScopes callback, i.e. how can I "attempt to use the same account in a new tab/client" in the test?

Reply:
Following [this conversation with CHS](https://magnopus.slack.com/archives/C02T07E8P2B/p1734365278798399), it turns out that the event is not currently being fired at all. It will be fired, after [this PR](https://github.com/magnopus/terraform-env-olympus2-development/pull/104) is merged, when:
- a user is removed from a Space 
- a space is deleted
- you connect to any space but haven't disconnected from previous sessions